ext-src/liveShare.ts

```diff
@@ -0,0 +1,136 @@
+import * as vsls from 'vsls/vscode';
+import { BrowserViewWindowManager, BrowserViewWindow, PANEL_TITLE } from './extension';
+
+const SERVICE_NAME = 'browser-preview';
+const REQUEST_GET_WINDOWS = 'getWindows';
+const NOTIFICATION_WINDOW_CREATED = 'windowCreated';
+const NOTIFICATION_WINDOW_INTERACTION = 'windowInteraction';
+
+const DISPATCHED_EVENTS = ['Input.dispatchMouseEvent', 'Input.dispatchKeyEvent'];
+const SHARED_PANEL_TITLE = `${PANEL_TITLE} (Shared)`;
+
+export async function setupLiveShare(windowManager: BrowserViewWindowManager) {
+  const liveShare = await vsls.getApi();
+  if (!liveShare) {
+    // The user doesn't have the Live Share
+    // extension installed, so there's nothing
+    // further to initialize.
+    return;
+  }
+
+  await sessionStarted(liveShare);
+  setupServices(liveShare, windowManager);
+}
+
+async function setupServices(liveShare: vsls.LiveShare, windowManager: BrowserViewWindowManager) {
+  let service: vsls.SharedService | vsls.SharedServiceProxy;
+
+  if (liveShare.session.role === vsls.Role.Host) {
+    service = (await liveShare.shareService(SERVICE_NAME))!;
+
+    service.onRequest(REQUEST_GET_WINDOWS, () => {
+      const windows = Array.from(windowManager.openWindows);
+      return windows.map((window) => {
+        const viewport = window.browserPage!.page.viewport();
+
+        return {
+          startUrl: window.config.startUrl, // TODO: This needs to be the current URL
```

Comment:
In order to fully support the "late join" scenario, we need to be able to capture the current URL of each preview window (not just their start URL), as well as the scroll position.

Reply:
State is managed from the front-end app, where the VS Code extension acts as a back-end API providing a few configurations. See https://github.com/auchenberg/vscode-browser-preview/blob/master/src/App.tsx#L10